File: src/Blokemon.PackGen/Rendering/Svg.cs
using System.Globalization;

namespace Blokemon.PackGen.Rendering;

/// <summary>A gradient stop.</summary>
/// <param name="Offset">The stop position as a fraction of the gradient line.</param>
/// <param name="Colour">The stop colour.</param>
/// <param name="Alpha">The stop opacity.</param>
public readonly record struct Stop(double Offset, string Colour, double Alpha = 1d);

/// <summary>Primitives shared by every printed element.</summary>
public static class Svg
{
    /// <summary>Formats a number for a coordinate or a length.</summary>
    /// <param name="value">The number to format.</param>
    /// <returns>The formatted number.</returns>
    public static string N(double value) =>
        Math.Round(value, 3).ToString("0.###", CultureInfo.InvariantCulture);

    /// <summary>Escapes text for markup.</summary>
    /// <param name="value">The text to escape.</param>
    /// <returns>The escaped text.</returns>
    public static string Esc(string value) =>
        value
            .Replace("&", "&amp;", StringComparison.Ordinal)
            .Replace("<", "&lt;", StringComparison.Ordinal)
            .Replace(">", "&gt;", StringComparison.Ordinal)
            .Replace("\"", "&quot;", StringComparison.Ordinal)
            .Replace("'", "&#39;", StringComparison.Ordinal);

    /// <summary>Defines a linear gradient across a box at a CSS gradient angle.</summary>
    /// <param name="id">The gradient identity.</param>
    /// <param name="angle">The CSS angle in degrees, measured clockwise from upwards.</param>
    /// <param name="width">The box width.</param>
    /// <param name="height">The box height.</param>
    /// <param name="stops">The stops along the gradient line.</param>
    /// <returns>The gradient definition.</returns>
    public static string Linear(
        string id,
        double angle,
        double width,
        double height,
        params Stop[] stops
    )
    {
        // CSS measures the gradient line clockwise from upwards and sizes it so the box corners
        // land exactly on its ends, which is why the length is a sum of projections rather than
        // the box diagonal.
        var radians = angle * Math.PI / 180d;
        var (dx, dy) = (Math.Sin(radians), -Math.Cos(radians));
        var length = (Math.Abs(width * dx) + Math.Abs(height * dy)) / 2d;
        var (cx, cy) = (width / 2d, height / 2d);

        var marks = string.Concat(
            stops.Select(stop =>
                $"""<stop offset="{N(stop.Offset)}" stop-color="{stop.Colour}" stop-opacity="{N(stop.Alpha)}"/>"""
            )
        );

        return $"""<linearGradient id="{id}" gradientUnits="userSpaceOnUse" x1="{N(cx - (dx * length))}" y1="{N(cy - (dy * length))}" x2="{N(cx + (dx * length))}" y2="{N(cy + (dy * length))}">{marks}</linearGradient>""";
    }

    /// <summary>Defines a stripe pattern running at a CSS gradient angle.</summary>
    /// <param name="id">The pattern identity.</param>
    /// <param name="angle">The CSS angle of the gradient the stripes run across.</param>
    /// <param name="period">The distance between stripes.</param>
    /// <param name="thickness">The stripe thickness.</param>
    /// <param name="colour">The stripe colour.</param>
    /// <param name="alpha">The stripe opacity.</param>
    /// <returns>The pattern definition.</returns>
    public static string Stripes(
        string id,
        double angle,
        double period,
        double thickness,
        string colour,
        double alpha
    )
    {
        // The stripes lie perpendicular to the gradient line, so a tile of vertical bars is
        // turned by the angle the gradient makes with the horizontal.
        var turn = angle - 90d;

        return $"""<pattern id="{id}" width="{N(period)}" height="{N(period)}" patternUnits="userSpaceOnUse" patternTransform="rotate({N(turn)})"><rect width="{N(thickness)}" height="{N(period)}" fill="{colour}" fill-opacity="{N(alpha)}"/></pattern>""";
    }

    /// <summary>A rectangle filled by a definition.</summary>
    /// <param name="x">The left edge.</param>
    /// <param name="y">The top edge.</param>
    /// <param name="width">The width.</param>
    /// <param name="height">The height.</param>
    /// <param name="fill">The fill.</param>
    /// <param name="extra">Attributes appended to the element.</param>
    /// <returns>The rectangle.</returns>
    public static string Rect(
        double x,
        double y,
        double width,
        double height,
        string fill,
        string extra = ""
    ) =>
        $"""<rect x="{N(x)}" y="{N(y)}" width="{N(width)}" height="{N(height)}" fill="{fill}"{extra}/>""";

    /// <summary>A reference to a definition.</summary>
    /// <param name="id">The definition identity.</param>
    /// <returns>The reference.</returns>
    public static string Url(string id) => $"url(#{id})";
}

File: src/Blokemon.PackGen/Rendering/Svg.fs
namespace Blokemon.PackGen.Rendering

open System
open System.Globalization

/// A gradient stop.
type Stop =
    {
        /// The stop position as a fraction of the gradient line.
        Offset: float

        /// The stop colour.
        Colour: string

        /// The stop opacity.
        Alpha: float
    }

/// Gradient stops.
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module Stop =

    /// A fully opaque stop.
    let solid offset colour =
        { Offset = offset
          Colour = colour
          Alpha = 1.0 }

    /// A stop at a declared opacity.
    let faded offset colour alpha =
        { Offset = offset
          Colour = colour
          Alpha = alpha }

/// Primitives shared by every printed element.
module Svg =

    /// Formats a number for a coordinate or a length.
    let n (value: float) =
        Math.Round(value, 3).ToString("0.###", CultureInfo.InvariantCulture)

    /// Escapes text for markup.
    let esc (value: string) =
        value
            .Replace("&", "&amp;")
            .Replace("<", "&lt;")
            .Replace(">", "&gt;")
            .Replace("\"", "&quot;")
            .Replace("'", "&#39;")

    /// Defines a linear gradient across a box at a CSS gradient angle, measured clockwise from
    /// upwards.
    let linear (id: string) (angle: float) (width: float) (height: float) (stops: Stop list) =
        // CSS measures the gradient line clockwise from upwards and sizes it so the box corners
        // land exactly on its ends, which is why the length is a sum of projections rather than
        // the box diagonal.
        let radians = angle * Math.PI / 180.0
        let dx, dy = sin radians, -cos radians
        let length = (abs (width * dx) + abs (height * dy)) / 2.0
        let cx, cy = width / 2.0, height / 2.0

        let marks =
            stops
            |> List.map (fun stop ->
                $"""<stop offset="{n stop.Offset}" stop-color="{stop.Colour}" stop-opacity="{n stop.Alpha}"/>""")
            |> String.concat ""

        $"""<linearGradient id="{id}" gradientUnits="userSpaceOnUse" x1="{n (cx - dx * length)}" y1="{n (cy - dy * length)}" x2="{n (cx + dx * length)}" y2="{n (cy + dy * length)}">{marks}</linearGradient>"""

    /// Defines a stripe pattern running across a gradient at a CSS gradient angle.
    let stripes
        (id: string)
        (angle: float)
        (period: float)
        (thickness: float)
        (colour: string)
        (alpha: float)
        =
        // The stripes lie perpendicular to the gradient line, so a tile of vertical bars is
        // turned by the angle the gradient makes with the horizontal.
        let turn = angle - 90.0

        $"""<pattern id="{id}" width="{n period}" height="{n period}" patternUnits="userSpaceOnUse" patternTransform="rotate({n turn})"><rect width="{n thickness}" height="{n period}" fill="{colour}" fill-opacity="{n alpha}"/></pattern>"""

    /// A rectangle filled by a definition, carrying attributes appended to the element.
    let rectWith
        (x: float)
        (y: float)
        (width: float)
        (height: float)
        (fill: string)
        (extra: string)
        =
        $"""<rect x="{n x}" y="{n y}" width="{n width}" height="{n height}" fill="{fill}"{extra}/>"""

    /// A rectangle filled by a definition.
    let rect x y width height fill = rectWith x y width height fill ""

    /// A reference to a definition.
    let url (id: string) = $"url(#{id})"
